src/legacy/core_plugins/embeddable_api/public/triggers/attach_action.ts

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { IRegistry, Trigger } from '../types';
+
+export function attachAction(
+  triggerRegistry: IRegistry<Trigger>,
+  { triggerId, actionId }: { triggerId: string; actionId: string }
```

Comment:
should we rather pass an action instead of actionId ? this way we can be sure action exists, and this can work without having an action in registry.

Reply:
We could possibly do that.  This was initially done with an id mapping because I was going to store them in the database, so when end users create trigger id -> action mappings, it's stored somewhere.  

Late in my exploration of phase 2 I realized we can't store these connections in a new saved object in the kibana index, and instead it'll have to be stored as information on the embeddable, so something like:

```
 const input = {
   dynamicActions: [{ 
     triggerId: APPLY_FILTER_TRIGGER,
     actionType: SAVED_OBJECT_DRILLDOWN_ACTION,
     actionConfiguration: "{objectType: 'dashboard', objectId: '123', customTimeRange: '...
', applyDynamicFilters: true}"
     actionId: '123444' // I think we will need an id to let the end user delete actions.
   }],
   viewMode: ...,
  ... other standard input parameters
 };
 new VisualizeEmbeddable(input, ...);
```

So that's the history of why I did it with the mapping of id to id... but given the turn of events at the end of the phase 2 exploration... now I think we could do your suggestion that when an embeddable is loaded up, it creates those action instances and then you could throw them on the triggerRegistry.

That being said I don't really think it's that much of a benefit.  You will still have to handle the possibility of an action no longer being able to be created into an instance.  e.g. a plugin creates a new action type (like SAVED_OBJECT_DRILLDOWN_ACTION, it'll be a commercial feature likely), an end user configures one for their embeddable, then they drop down a license, or someone else imports a dashboard but the plugin that knows how to instantiate a new instance of SAVED_OBJECT_DRILLDOWN_ACTION doesn't exist, so at some point we will still have to handle this kind of error.  

The other benefit of having it mapped to id is that you can more easily see if you are adding a duplicate trigger -> same exact action.  Another benefit is that I believe with phase 2 you can take up less memory space, by creating the instance of the action only at the time the trigger is executed, rather than keeping the entire array of instances in memory.

I'm open to more discussion though if you still feel like `trigger id -> action instance` mapping is better than `trigger id -> action id` mapping.